src/Cognito.re

```diff
@@ -201,6 +206,158 @@ let signUp =
           | _ => `CognitoUnknownError(msg)
           };
         Future.make(resolve => resolve(Belt.Result.Error(err)));
-      }
-    );
+      };
+    });
+};
+type confirmSignUpErrors = [ | `CognitoConfirmationCodeValidation(string)];
+let confirmSignUp = (config, ~username, ~confirmationCode, ()) => {
+  let params = Js.Dict.empty();
+
+  Js.Dict.set(params, "Username", Js.Json.string(username));
+  Js.Dict.set(params, "ConfirmationCode", Js.Json.string(confirmationCode));
+
+  Client.request(config, ConfirmSignUp, params)
+  ->Future.flatMapOk(res => {
+      /*
+             This is the raw response. Not sure what to do with this if anything. What is `tag:1`
+             Raw Response [ [ 200, tag: 1 ], {} ]
+       Raw Response [ [ 200, tag: 1 ], {} ] */
+      Js.log2("Raw Response", res);
+
+      switch (res.status) {
+      | Success(_) =>
+        // We're _really_ hoping amazon holds to their API contract here.
+        // If not, we're going to see null type errors.
+        Future.make(resolve =>
+          resolve(
+            Belt.Result.Ok(
+              {
+                res;
+              },
+            ),
+          )
+        )
+      | Informational(_)
+      | Redirect(_)
+      | ClientError(_)
+      | ServerError(_) =>
+        // We're _really_ hoping amazon holds to their API contract here too.
+        // Although the unknownerror variant helps lots.
+        let isErrorResponse = makeResponse(res.json);
+        let err = isErrorResponse->__typeGet;
+        let msg = isErrorResponse->messageGet;
+        let err =
+          switch (err) {
+          | "CodeMismatchException" => `CognitoConfirmationCodeValidation(msg)
```

Comment:
```suggestion
          | "CodeMismatchException" => `CognitoCodeMismatch(msg)
```
Let's keep the terminology as close to AWS' as possible while stilll being valid reason.  I did drop the extraneous word "exception" here since this is a Result.Error() so we know we should be handling it/ignoring it appropriately.

Reply:
See above, there's a lot more errors documented by AWS at that link, we should handle them all here.  There should be some common cases from SignUp, so feel free to just copy that code and we can worry about an abstraction later.